### DISCUSSION TITLE
AWS authentication through STS AssumeRole API

### DISCUSSION THREAD

**Discussion by antoineco** [outline/outline#5923] [Ideas]
As of v0.72.0, Outline only supports authentication to Amazon S3 using AWS security credentials (access key):

https://github.com/outline/outline/blob/56cae8a54563c0cd3b2dd28d64baa184a45c490f/server/storage/files/S3Storage.ts#L16-L17

Applications which run inside AWS—like on an EC2 instance or inside an EKS cluster—can typically [assume an IAM role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html) instead of requiring explicit credentials.

Most AWS SDKs expose some factory function that can automatically configure a suitable credentials provider to be used by AWS clients. For NodeJS, this function is [`fromNodeProviderChain()`](https://www.npmjs.com/package/@aws-sdk/credential-providers/v/3.423.0#fromnodeproviderchain) (uses [@aws-sdk/credential-provider-node](https://www.npmjs.com/package/@aws-sdk/credential-provider-node/v/3.423.0) under the hood).

```js
const { fromNodeProviderChain } = require("@aws-sdk/credential-providers");
const credentials = fromNodeProviderChain();
const client = new S3Client({ credentials });
```

_In fact, all Node.js AWS clients use this provider by default, in absence of explicitly provided alternative._

The above example is for the v3 version of the SDK—which requires outline/outline#5683—but from what I observed simply omitting the `accessKeyId` and `secretAccessKey` attributes unset has the same result in v2 (with the added benefit that `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` are still being honored implicitly, if set).

**Comment by tommoor** [comment]
What is the behavior that you are seeing? This shouldn’t have changed

**Reply by antoineco** [reply]
Nothing has changed, Outline just never supported this, hence the feature request.

**Reply by tommoor** [reply]
You said "As of v0.72.0" – so it sounded like something had explicitly changed in yesterdays release.

I know for a fact that several customers use assumed IAM roles instead of keys, there shouldn't be anything stopping it, it's just not documented.

**Reply by antoineco** [reply]
Oh, sorry for the confusion! I mentioned the current version as a habit, otherwise it can get difficult to relate a certain feature request to the context at the time the discussion (or issue) was opened.

---

Thanks for confirming that this scenario is supported. I indeed managed to make it work 👍

For more context, I had been observing failures to upload attachments to S3, and been assuming that authentication was the cause because AWS was showing "Last activity: never" on the IAM role associated with Outline.
It turns out that the error was client-side, not server-side:

```
Access to XMLHttpRequest at 'https://my-bucket.s3.eu-fake-0.amazonaws.com/' from origin 'https://wiki.example.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

After fixing my [allowed origins](https://docs.getoutline.com/s/125de1cc-9ff6-424b-8415-0d58c809a40f#h-aws-s3) on the S3 bucket, attachments started working.
